### ISSUE TITLE
Type mismatch with async block to `dyn Future` coercion

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#11815]
```rs
use core::future::Future;
use core::pin::Pin;

// expected Pin<Box<dyn Future<Output = ()> + Send, Global>>, found Pin<Box<impl Future<Output = ()>, Global>>
fn send() -> Pin<Box<dyn Future<Output = ()> + Send + 'static>> {
    Box::pin(async move {})
}

// expected Pin<Box<dyn Future<Output = ()>, Global>>, found Pin<Box<impl Future<Output = ()>, Global>>
fn not_send() -> Pin<Box<dyn Future<Output = ()> + 'static>> {
    Box::pin(async move {})
}
```
high priority as this affects all uses of `async_trait`

**Comment by kpreid** [comment]
Possibly-the-same-bug appears with `async` and `Future`: this code
```rust
fn foo() {
    let a = noise::Value::new();
    let b: noise::ScaleBias<'_, [f64; 2]> = noise::ScaleBias::new(&a).set_bias(1.0);
}
```
(with [`noise = "0.7.0"`](https://crates.io/crates/noise/0.7.0)) produces the false error `expected &dyn NoiseFn<[f64; 2]>, found &Value`. I haven't found a way to reproduce it with a simple locally-defined trait, though.

**Comment by Emilgardis** [comment]
is there a way to locally disable the check for cases where this hits?

**Comment by lnicola** [comment]
@Emilgardis not right now. Only per-project.

**Comment by BlackDex** [comment]
Isn't this fixed already?
The current latest nightly version works for me without any issues.

**Comment by flodiebold** [comment]
The error mentioned in the issue is still showing up, and so is the `noise` error.

**Comment by arilotter** [comment]
Any pointers on getting started on this bug? It's a major blocker for my Rust development use case - I can't use RA at all for work until this is fixed, and have to rely on `cargo watch`. Would love to help fix it.

**Comment by lnicola** [comment]
@arilotter this diagnostic is opt-in in newer RA versions.

**Comment by kpreid** [comment]
@lnicola Where is that controlled? I'm still getting the false error and I've checked my user and workspace VSCode `settings.json` for rust-analyzer settings that sound related to this, and I don't see any.

**Comment by flodiebold** [comment]
It's `rust-analyzer.diagnostics.enableExperimental`.

**Comment by kpreid** [comment]
I don't have that setting enabled. I'm on r-a extension `v0.2.1040` server `5dce1ff02 2022-05-02 stable`, and my settings files contain only the following r-a settings:

```json
{
    "rust-analyzer.inlayHints.parameterHints": false,
    "rust-analyzer.inlayHints.maxLength": 5,
    "rust-analyzer.inlayHints.typeHints": false,
    "rust-analyzer.diagnostics.disabled": [
        "missing-unsafe"
    ],
    "rust-analyzer.inlayHints.enable": false
}
```

```json
{
    "rust-analyzer.cargo.allFeatures": true,
    "rust-analyzer.assist.importPrefix": "crate",
    "rust-analyzer.checkOnSave.command": "clippy",
    "rust-analyzer.checkOnSave.allTargets": true,
    "rust-analyzer.assist.importGranularity": "module"
}
```

**Comment by arilotter** [comment]
`rust-analyzer.diagnostics.enableExperimental` was on by default for me.

**Comment by flodiebold** [comment]
> Any pointers on getting started on this bug? Would love to help fix it.

I think the basic problem here is that currently, we just model async blocks as an opaque type, _and_ we don't implement any logic for the hidden type behind the opaque type:
https://github.com/rust-lang/rust-analyzer/blob/1f709d54463972b189a3120be4073c507f2fbc00/crates/hir-ty/src/chalk_db.rs#L282-L285
As the FIXME says, the hidden type is used to determine auto traits, and this is probably the reason Chalk says the coercion doesn't work because it doesn't implement `Send`.
To really properly handle this, we'd need to desugar the async block to a generator like rustc does, _and_ actually implement generators. Or at least give it some generator type, and make sure the generator type implements all necessary auto traits somehow. Alternatively we could do some hack to consider these opaque types to always implement auto traits, though I'm not sure what the best way to do this is.

**Comment by flodiebold** [comment]
This is fixed.
